src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseStringContainsCharOverloadWithSingleCharactersAnalyzer.cs

```diff
@@ -0,0 +1,97 @@
+            CA1841,
+            s_localizableTitle_CA1841,
+            s_localizableMessage_CA1841,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            description: s_localizableDescription_CA1841,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_rule_CA1841);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(CheckIfRuleIsApplicableAndRegister);
+        }
+
+        private void CheckIfRuleIsApplicableAndRegister(CompilationStartAnalysisContext context)
+        {
+            var stringType = context.Compilation.GetSpecialType(SpecialType.System_String);
+            var charType = context.Compilation.GetSpecialType(SpecialType.System_Char);
+
+            // Bail out of further evaluations in special cases where string/char types arent even available
+            if (stringType is null || charType is null)
+                return;
+
+            var ruleIsApplicable = stringType.GetMembers("Contains").OfType<IMethodSymbol>().Any(m =>
+            {
+                if (m.Parameters.Length > 0)
+                {
+                    return m.Parameters[0].Type.SpecialType == SpecialType.System_Char;
+                }
+                return false;
+            });
+
+            if (ruleIsApplicable)
+                context.RegisterOperationAction(AnalyseContainsUsage, OperationKind.Invocation);
+        }
+
+        private void AnalyseContainsUsage(OperationAnalysisContext context)
+        {
+            var invocationOperation = (IInvocationOperation)context.Operation;
+            if (ShouldAnalyzeInvocations(invocationOperation) && CheckForViolationAndTransformLiteral(invocationOperation.Arguments[0], out var sourceString))
+            {
+                var propertyBuilder = ImmutableDictionary.CreateBuilder<string, string?>(StringComparer.Ordinal);
+                propertyBuilder.Add(SourceStringPropertyKey, sourceString);
+                context.ReportDiagnostic(invocationOperation.Arguments[0].CreateDiagnostic(s_rule_CA1841, propertyBuilder.ToImmutable()));
+            }
+
+            static bool ShouldAnalyzeInvocations(IInvocationOperation invocationOperation)
+            {
+                return invocationOperation.TargetMethod is IMethodSymbol invokedMethod
+                                && invokedMethod.ContainingType.SpecialType.Equals(SpecialType.System_String)
+                                && invokedMethod.Name == "Contains"
+                                && invokedMethod.Parameters[0].Type.SpecialType.Equals(SpecialType.System_String);
```

Comment:
This will throw an exception if the `Contains` method has no parameters (possible for say a malformed or user-defined corlib). I think we should add a defensive length check.
```suggestion
                                && invokedMethod.Parameters.Length > 0
                                && invokedMethod.Parameters[0].Type.SpecialType.Equals(SpecialType.System_String);
```